Add api for recursively analyzing function calls

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73329
* #73270
* #69871

There is a quantization use case for having better alias analysis with function calls remaining. This does the relatively dumb approach of getting the inlined graph of each function call, and then analyzing that subgraph. Since we need a unique single analysis of every `Value*`, for every function call make a copy of the graph for every analysis past the first. This is relatively slow, but given the limited use case here should work well enough (and is no slower than calling the inlining pass). 

cc @vkuzo

Differential Revision: [D34451424](https://our.internmc.facebook.com/intern/diff/D34451424)

Commit:
Add api for recursively analyzing function calls

[ghstack-poisoned]

Commit:
Update on "Add api for recursively analyzing function calls"




[ghstack-poisoned]

Commit:
Update on "Add api for recursively analyzing function calls"


There is a quantization use case for having better alias analysis with function calls remaining. This does the relatively dumb approach of getting the inlined graph of each function call, and then analyzing that subgraph. Since we need a unique single analysis of every `Value*`, for every function call make a copy of the graph for every analysis past the first. This is relatively slow, but given the limited use case here should work relatively well enough (and is no slower than calling the inlining pass). 

cc vkuzo 

[ghstack-poisoned]

Commit:
Update on "Add api for recursively analyzing function calls"


There is a quantization use case for having better alias analysis with function calls remaining. This does the relatively dumb approach of getting the inlined graph of each function call, and then analyzing that subgraph. Since we need a unique single analysis of every `Value*`, for every function call make a copy of the graph for every analysis past the first. This is relatively slow, but given the limited use case here should work well enough (and is no slower than calling the inlining pass). 

cc vkuzo 

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34451424).

davidberard98 commented on torch/csrc/jit/passes/inliner.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <torch/csrc/jit/api/module.h>
 #include <torch/csrc/jit/frontend/error_report.h>
 #include <torch/csrc/jit/jit_log.h>
+#include <ATen/core/interned_strings.h>
```
nit: linter error here in case you missed it (wants you to move it to line 3 )

davidberard98 commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -947,11 +986,15 @@ void AliasDb::analyzeSubgraph(Node* node) {
   TORCH_INTERNAL_ASSERT(
       subgraphBlock->outputs().size() >= node->outputs().size());
-  for (const auto i : c10::irange(node->outputs().size())) {
+  for (size_t i = 0; i < node->outputs().size(); i++) {
```
nit: any reason not to use irange?

eellison replied:
there are too many C++ idioms to remember 😢

davidberard98 reviewed (approved):
Looks good!

Gamrix reviewed (approved):
Only thing would be adding the check that objects from separate function calls aren't aliasing each other.

Commit:
Update on "Add api for recursively analyzing function calls"


There is a quantization use case for having better alias analysis with function calls remaining. This does the relatively dumb approach of getting the inlined graph of each function call, and then analyzing that subgraph. Since we need a unique single analysis of every `Value*`, for every function call make a copy of the graph for every analysis past the first. This is relatively slow, but given the limited use case here should work well enough (and is no slower than calling the inlining pass). 

cc vkuzo

Differential Revision: [D34451424](https://our.internmc.facebook.com/intern/diff/D34451424)

[ghstack-poisoned]

Commit:
Update on "Add api for recursively analyzing function calls"


There is a quantization use case for having better alias analysis with function calls remaining. This does the relatively dumb approach of getting the inlined graph of each function call, and then analyzing that subgraph. Since we need a unique single analysis of every `Value*`, for every function call make a copy of the graph for every analysis past the first. This is relatively slow, but given the limited use case here should work well enough (and is no slower than calling the inlining pass). 

cc vkuzo

Differential Revision: [D34451424](https://our.internmc.facebook.com/intern/diff/D34451424)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34451424).

Commit:
Update on "Add api for recursively analyzing function calls"


There is a quantization use case for having better alias analysis with function calls remaining. This does the relatively dumb approach of getting the inlined graph of each function call, and then analyzing that subgraph. Since we need a unique single analysis of every `Value*`, for every function call make a copy of the graph for every analysis past the first. This is relatively slow, but given the limited use case here should work well enough (and is no slower than calling the inlining pass). 

cc vkuzo

Differential Revision: [D34451424](https://our.internmc.facebook.com/intern/diff/D34451424)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34451424).

swolchok commented on torch/csrc/jit/passes/inliner.cpp:
```diff
@@ -12,18 +13,34 @@ namespace prim {
+  if (n->kind() == prim::CallFunction) {
+    AT_ASSERT(n->input(0)->node()->kind() == prim::Constant);
+    auto function_constant = n->input(0)->node();
+    auto fun_type = function_constant->output()->type()->expect<FunctionType>();
```
use expectRef() -- `const auto& fun_type = function_constant->output()->type()->expectRef<FunctionType>()`

eellison replied:
this is already the existing status code.. will fix in follow diff trying to land this today.

swolchok commented on torch/csrc/jit/passes/inliner.cpp:
```diff
@@ -12,18 +13,34 @@ namespace prim {
+  }
+  if (n->kind() == prim::CallMethod) {
+    const std::string& name = n->s(attr::name);
+    if (auto class_type = n->input(0)->type()->cast<ClassType>()) {
```
castRaw

swolchok commented on torch/csrc/jit/passes/inliner.cpp:
```diff
@@ -12,18 +13,34 @@ namespace prim {
-            function_constant->output()->type()->expect<FunctionType>();
+        if (auto graphFunction = tryToGraphFunction(cur)) {
+          auto function_constant = cur->input(0)->node();
+          auto fun_type =
```
expectRef

swolchok reviewed (commented):
probably shouldn't be reviewing for actual content, but here are some minor efficiency complaints

Commit:
Update on "Add api for recursively analyzing function calls"


There is a quantization use case for having better alias analysis with function calls remaining. This does the relatively dumb approach of getting the inlined graph of each function call, and then analyzing that subgraph. Since we need a unique single analysis of every `Value*`, for every function call make a copy of the graph for every analysis past the first. This is relatively slow, but given the limited use case here should work well enough (and is no slower than calling the inlining pass). 

cc vkuzo

Differential Revision: [D34451424](https://our.internmc.facebook.com/intern/diff/D34451424)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34451424).

Commit:
Update on "Add api for recursively analyzing function calls"


There is a quantization use case for having better alias analysis with function calls remaining. This does the relatively dumb approach of getting the inlined graph of each function call, and then analyzing that subgraph. Since we need a unique single analysis of every `Value*`, for every function call make a copy of the graph for every analysis past the first. This is relatively slow, but given the limited use case here should work well enough (and is no slower than calling the inlining pass). 

cc vkuzo

Differential Revision: [D34451424](https://our.internmc.facebook.com/intern/diff/D34451424)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34451424).

Commit:
Update on "Add api for recursively analyzing function calls"


There is a quantization use case for having better alias analysis with function calls remaining. This does the relatively dumb approach of getting the inlined graph of each function call, and then analyzing that subgraph. Since we need a unique single analysis of every `Value*`, for every function call make a copy of the graph for every analysis past the first. This is relatively slow, but given the limited use case here should work well enough (and is no slower than calling the inlining pass). 

cc vkuzo

Differential Revision: [D34451424](https://our.internmc.facebook.com/intern/diff/D34451424)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34451424).
